### PR TITLE
PERL-792 Retryable Writes

### PR DESCRIPTION
**Pull request by TBSliver**

### CODE REVIEW

File: devel/t-dynamic/PERL-792-replica-set-failover.t [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+my $post_stepdown_ret = $coll->insert_one( { _id => 2, test => 'again' } );
+
+is $post_stepdown_ret->inserted_id, 2, 'write succeeded';
+
+# All this is to make sure we dont make assumptions on position of the actual
+# event, just that the failed one comes first.
+my $first_insert_index;
+
+for my $f_idx ( 0 .. $#events ) {
```
Comment (xdg) [inline devel/t-dynamic/PERL-792-replica-set-failover.t]: `$#events -1`?  Or line 199 could go off the end.  Should never happen if command monitoring is working right and always pairing, but I'd rather not rely on a reader knowing that.

File: devel/t-dynamic/PERL-792-replica-set-failover.t [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        $first_insert_index = $f_idx;
+        last;
+    }
+}
+
+ok defined( $first_insert_index ), 'found first command';
+
+my $second_insert_index;
+
+for my $s_idx ( $first_insert_index + 2 .. $#events ) {
```
Comment (xdg) [inline devel/t-dynamic/PERL-792-replica-set-failover.t]: Ditto, plus it's worth checking there are enough events past the first.

File: lib/MongoDB/Op/_BulkWrite.pm [hunk 3]
```diff
@@ -74,6 +87,46 @@ sub has_collation {
     } @{ $self->queue };
 }
 
+## Encapsulate the write loop for this
```
Comment (xdg) [inline lib/MongoDB/Op/_BulkWrite.pm]: Please figure out what of this comment needs to be kept (if any) for documentation purposes and what can be deleted.

File: lib/MongoDB/Op/_BulkWrite.pm [hunk 4]
```diff
@@ -223,9 +278,10 @@ sub _execute_write_command_batch {
                     );
                 }
                 else {
-                    unshift @left_to_send, $self->_split_chunk( $link, $chunk, $_->size );
+                    unshift @left_to_send, $self->_split_chunk( $chunk, $_->size );
                 }
             }
+            # Put retryable writes catches... no not here... damn... ermm...  oh. Could increment transaction id at the beginning before the loop, then after the redo increment again. Then it wont get caught? hmm.... food
```
Comment (xdg) [inline lib/MongoDB/Op/_BulkWrite.pm]: ?!?

File: lib/MongoDB/Role/_RetryableBulk.pm [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+use strict;
+use warnings;
+package MongoDB::Role::_RetryableBulk;
```
Comment (xdg) [inline lib/MongoDB/Role/_RetryableBulk.pm]: Having a role just to provide an attribute seems excessive.  I think it's OK to just put it in the two classes that need it.

File: t/PERL-792-command-construction.t [hunk 6]
```diff
@@ -0,0 +1,320 @@
+#
```
Comment (xdg) [inline t/PERL-792-command-construction.t]: I like how this file turned out!  Very easy to follow.

File: lib/MongoDB/Op/_BulkWrite.pm (hunk 4 above)
Comment (TBSliver) [reply lib/MongoDB/Op/_BulkWrite.pm]: Ah oops I thought i got rid of that comment... Was from the odd debugging stuff when I tied the driver in knots with $link

**Comment by TBSliver** [comment]
there we go! Hope that all works :D

**Comment by xdg** [comment]
LGTM. I'll try to rebase and get it merged.

**Comment by xdg** [comment]
Rebased, fixed up, and merged.  Thanks!

**Pull request closed without merging**